larcorealg/Geometry/OpDetGeo.h

```diff
@@ -139,15 +145,60 @@ namespace geo {
     /// Returns the ROOT object describing the detector geometry.
     const TGeoNode*     Node() const { return fOpDetNode; }
 
+    
+    // --- BEGIN -- detector shape ---------------------------------------------
+    /// @name Detector shape
+    /// @{
+    
     /// Returns the geometry object as `TGeoShape`.
     TGeoShape const* Shape() const { return Node()->GetVolume()->GetShape(); }
 
-    /// Returns whether the detector shape is a cilynder (`TGeoTube`).
-    bool isTube() const { return asTube() != nullptr; }
+    /**
+     * @brief Returns whether the detector has the specified shape.
+     * @tparam ShapeObj type of ROOT geometry object representing the shape
+     * @return whether this detector has the specified shape
+     * @see `isShapeLike()`, `isBox()`, `isSphere()`, `isTube()`
+     * 
+     * Example:
+     * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+     * bool const isSphere = opDet.isShape<TGeoSphere>();
+     * bool const isBox = opDet.isShape<TGeoBBox>();
+     * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+     * will have `isSphere` `true` only if the shape of this object is a sphere
+     * (`TGeoSphere`), and `isBox` `true` only if the shape of this object is a
+     * box (`TGeoBBox`).
+     */
+    template <typename ShapeObj>
+    bool isShape() const;
+    
+    /**
+     * @brief Returns whether the detector inherits from the specified shape.
+     * @tparam ShapeObj type of ROOT geometry object representing the shape
+     * @return whether this detector has a shape derived from the specified one
+     * @see `isShape()`, `isBox()`, `isSphere()`, `isTube()`
+     * 
+     * Example:
+     * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+     * bool const isTubeLike = opDet.isShapeLike<TGeoTube>();
+     * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+     * `isTubeLike` will be `true` if its shape is either a box (`TGeoTube`)
+     * or any other whose shape object is derived from `TGeoTube` (including
+     * for example a C-shape, half-cylinder).
+     */
+    template <typename ShapeObj>
+    bool isShapeLike() const;
+    
+    /// Returns whether the detector shape is a cylinder (`TGeoTube`).
+    bool isTube() const { return isShapeLike<TGeoTube>(); }
 
     /// Returns whether the detector shape is a bar (`TGeoBBox`).
-    bool isBar() const { return (asBox() != nullptr) && !isTube(); }
+    bool isBar() const { return isShape<TGeoBBox>(); }
```

Comment:
This is not the correct implementation--`TSphere` inherits from `TGeoBBox` (bounding box) so it will return `true` when it should not.  The implementation you're replacing is the correct one.

Reply:
`isShape()` is supposed to match the type exactly, not including base classes.
"Unfortunately", I made a mistake and I was comparing the type of the requested class with the type of _a pointer_ to a class and, guess what? that test _always_ failed.
My testing with `icaruscode` did not discover the issue for two unfortunate coincident factors (ICARUS has _only_ spheres, and spheres are treated slightly different than other objects in the dump, skipping `isSphere()` and `isShape()`).
I have now pulled in `sbndcode` (which I had problem to convince CMake to process, by the way), and I think now it is correct; please check it anyway if you can!